presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/LogicalCteOptimizer.java

```diff
@@ -0,0 +1,197 @@
+            if (topologicalOrderedList.isEmpty()) {
+                return root;
+            }
+            SequenceNode sequenceNode = new SequenceNode(root.getSourceLocation(), planNodeIdAllocator.getNextId(), topologicalOrderedList,
+                    transformedCte.getSources().get(0));
+            return root.replaceChildren(Arrays.asList(sequenceNode));
+        }
+    }
+
+    public class CteConsumerTransformer
+            extends SimplePlanRewriter<CteTransformerContext>
+    {
+        private final PlanNodeIdAllocator idAllocator;
+
+        private final VariableAllocator variableAllocator;
+
+        public CteConsumerTransformer(PlanNodeIdAllocator idAllocator, VariableAllocator variableAllocator)
+        {
+            this.idAllocator = idAllocator;
+            this.variableAllocator = variableAllocator;
+        }
+
+        @Override
+        public PlanNode visitCteReference(CteReferenceNode node, RewriteContext<CteTransformerContext> context)
+        {
+            context.get().addDependency(node.getCteName());
+            context.get().setActiveCte(node.getCteName());
+            // So that dependent CTEs are processed properly
+            PlanNode actualSource = context.rewrite(node.getSource(), context.get());
+            context.get().clearActiveCte();
+            CteProducerNode cteProducerSource = new CteProducerNode(node.getSourceLocation(),
+                    idAllocator.getNextId(),
+                    actualSource,
+                    node.getCteName(),
+                    variableAllocator.newVariable("rows", BIGINT), node.getOutputVariables());
+            context.get().addProducer(node.getCteName(), cteProducerSource);
+            return new CteConsumerNode(node.getSourceLocation(), idAllocator.getNextId(), actualSource.getOutputVariables(), node.getCteName());
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<CteTransformerContext> context)
+        {
+            return new ApplyNode(node.getSourceLocation(),
+                    idAllocator.getNextId(),
+                    context.rewrite(node.getInput(),
+                            context.get()),
+                    context.rewrite(node.getSubquery(),
+                            context.get()),
+                    node.getSubqueryAssignments(),
+                    node.getCorrelation(),
+                    node.getOriginSubqueryError(),
+                    node.getMayParticipateInAntiJoin());
+        }}
+
+    public class CteTransformerContext
+    {
+        private static final String INACTIVE = "";
+        public Map<String, CteProducerNode> cteProducerMap;
+
+        // a -> b indicates that b needs to be processed before a
+        MutableGraph<String> graph;
+
+        public String activeCte;
+
+        public CteTransformerContext()
+        {
+            cteProducerMap = new HashMap<>();
+            // The cte graph will never have cycles because sql won't allow it
```

Comment:
I need to revert to my original dfs implementation because its not trivial to add a cycle check, since mutable graph expects the graph to be directed. But during review it was observed that that implementation added unnecessary complexity